protocol-units/dispute/src/MCR.sol

```diff
@@ -0,0 +1,165 @@
+    event OptimisticCommitmentSubmitted(bytes32 indexed blockHash, bytes stateCommitment, uint256 validatorStake);
+
+    constructor(
+        uint256 _delta,
+        uint256 _supermajorityStake,
+        uint256 _epochDurationInDays
+    ) {
+        delta = _delta;
+        supermajorityStake = _supermajorityStake;
+        epochDuration = _epochDurationInDays * SECONDS_IN_DAY;
+    }
+
+    function updateEpoch() public {
+        uint256 epochsPassed = (block.timestamp - epochStartTimestamp) / epochDuration;
+        currentEpoch += epochsPassed;
+        epochStartTimestamp += epochsPassed * epochDuration;
+    }
+
+    function stake() external payable {
+        validators[msg.sender].isRegistered = true;
+        validators[msg.sender].stake += msg.value;
+        totalStake += msg.value;
+        emit ValidatorRegistered(msg.sender, msg.value);
+    }
+
+    function unstake(uint256 amount) external {
+        Validator storage validator = validators[msg.sender];
+        require(validator.isRegistered, "Validator not registered");
+        require(validator.stake >= amount, "Insufficient stake");
+        validator.stake -= amount;
+        totalStake -= amount;
+        if (validator.stake == 0) {
+            validator.isRegistered = false;
+        }
+        payable(msg.sender).transfer(amount);
+        emit ValidatorDeregistered(msg.sender);
+    }
+
+    function getValidatorStatus() external view returns (bool, uint256) {
+        return (validators[msg.sender].isRegistered, validators[msg.sender].stake);
+    }
+
+    function submitDispute(bytes32 blockHash) external {
+        bytes32 disputeHash = keccak256(abi.encodePacked(blockHash, msg.sender, block.timestamp));
+        require(disputes[disputeHash].timestamp == 0, "Dispute already submitted");
+        disputes[disputeHash] = Dispute(blockHash, block.timestamp, DisputeState.SUBMITTED);
+        emit DisputeSubmitted(disputeHash, blockHash, msg.sender);
+    }
+
+    function resolveDispute(bytes32 disputeHash, DisputeState state) external {
+        require(state > DisputeState.SUBMITTED && state <= DisputeState.UNVERIFIABLE, "Invalid state transition");
+        Dispute storage dispute = disputes[disputeHash];
+        require(dispute.timestamp != 0, "Dispute not found");
+        require(dispute.state == DisputeState.SUBMITTED, "Dispute already resolved");
+        dispute.state = state;
+        emit DisputeResolved(disputeHash, state);
+    }
+
+    function isCommitmentAccepted(uint256 epoch) public view returns (bool) {
+        return epochs[epoch].isAccepted;
+    }
+
+    function submitOptimisticCommitment(bytes32 blockHash, bytes calldata stateCommitment) external {
+        require(validators[msg.sender].isRegistered, "Validator not registered");
+
+        updateEpoch();
+
+        OptimisticCommitment storage commitment = epochs[currentEpoch];
+        commitment.blockHash = blockHash;
```

Comment:
This field can be overwritten by any registered validator and once accepted, the last value is what will be emitted in `BlockAccepted`. Is this by design?

Reply:
I believe this renders the same race condition as in the original RStarM implementation. 

It should look more like this (pseudo-code from RFC)...

```
 function submitOptimisticCommitment(bytes32 blockHash, bytes calldata stateCommitment) external {

        require(validators[msg.sender].isRegistered, "Validator not registered");

        OptimisticCommitment storage commitment = optimisticCommitments[blockHash];
        
        // Increment the count for the submitted stateCommitment
        uint256 currentCount = commitment.stateCommitments[stateCommitment]++;
        // Update the state with the highest commit count if this one surpasses it
        if (currentCount + 1 > commitment.highestCommitCount) {
            commitment.highestCommitCount = currentCount + 1;
            commitment.highestCommitState = stateCommitment;
        }

        emit OptimisticCommitmentSubmitted(blockHash, stateCommitment, currentCount + 1);

        // Check if the count of the highest commitment has reached the threshold
        if (commitment.highestCommitCount >= m) {
            // Block is accepted optimistically after receiving minimum number of commitments
            emit BlockAccepted(blockHash, commitment.highestCommitState);
        }

    }
```

I will be deving on this shortly. Likely will be something I address.